pkg/store/bucket.go

```diff
@@ -1188,89 +1150,200 @@ func newBucketIndexReader(ctx context.Context, logger log.Logger, block *bucketB
+func (r *bucketIndexReader) fetchPostings(keys labels.Labels) (index.Postings, error) {
+	const maxGapSize = 512 * 1024
+
+	var (
+		ptrs     []postingPtr
+		postings = make([]index.Postings, 0, len(keys))
+	)
+
+	// TODO(bwplotka): sort postings?
+
+	for _, k := range keys {
+		// Get postings for given key from cache first.
+		if b, ok := r.cache.postings(r.block.meta.ULID, k); ok {
+			r.stats.postingsTouched++
+			r.stats.postingsTouchedSizeSum += len(b)
+
+			_, l, err := r.dec.Postings(b)
+			if err != nil {
+				return nil, errors.Wrap(err, "decode postings")
+			}
+			postings = append(postings, l)
+			continue
+		}
+
+		// Cache miss; save pointer for actual posting in index stored in object store.
+		ptr, ok := r.block.postings[k]
+		if !ok {
+			// Index malformed? Should not happen.
+			continue
 		}
-		p.set(l)
-		r.cache.setPostings(r.block.meta.ULID, p.key, c)
-		// If we just fetched it we still have to update the stats for touched postings.
-		r.stats.postingsTouched++
-		r.stats.postingsTouchedSizeSum += len(c)
+
+		ptrs = append(ptrs, postingPtr{ptr: ptr, key: k})
 	}
-	return nil
+
+	sort.Slice(ptrs, func(i, j int) bool {
+		return ptrs[i].ptr.Start < ptrs[j].ptr.Start
+	})
+
+	// TODO(bwplotka): Asses how large in worst case scenario this can be. (e.g fetch for AllPostingsKeys)
+	// Consider sub split if too big.
+	parts := partitionRanges(len(ptrs), func(i int) (start, end uint64) {
+		return uint64(ptrs[i].ptr.Start), uint64(ptrs[i].ptr.End)
+	}, maxGapSize)
+
+	var g run.Group
+	for _, p := range parts {
+		ctx, cancel := context.WithCancel(r.ctx)
+		i, j := p.elemRng[0], p.elemRng[1]
+
+		start := int64(p.start)
+		// We assume index does not have any ptrs that has 0 length.
+		lenght := int64(p.end) - start
+
+		// Fetch from object storage concurrently and update stats and posting list.
+		g.Add(func() error {
+			begin := time.Now()
+
+			b, err := r.block.readIndexRange(ctx, start, lenght)
+			if err != nil {
+				return errors.Wrap(err, "read postings range")
+			}
+			fetchTime := time.Since(begin)
+
+			r.mtx.Lock()
+			defer r.mtx.Unlock()
+
+			r.stats.postingsFetchCount++
+			r.stats.postingsFetched += len(ptrs)
+			r.stats.postingsFetchDurationSum += fetchTime
+			r.stats.postingsFetchedSizeSum += int(lenght)
+
+			for _, p := range ptrs[i:j] {
+				c := b[p.ptr.Start-start : p.ptr.End-start]
+
+				_, l, err := r.dec.Postings(c)
+				if err != nil {
+					return errors.Wrap(err, "read postings list")
+				}
+
+				// Return postings and fill LRU cache.
+				postings = append(postings, l)
+				r.cache.setPostings(r.block.meta.ULID, p.key, c)
+
+				// If we just fetched it we still have to update the stats for touched postings.
+				r.stats.postingsTouched++
+				r.stats.postingsTouchedSizeSum += len(c)
+			}
+			return nil
+		}, func(err error) {
+			if err != nil {
+				cancel()
+			}
+		})
+	}
+
+	if err := g.Run(); err != nil {
+		return nil, err
+	}
+
+	return index.Merge(postings...), nil
```

Comment:
would it be worth checking here if the returned can be cast to `ErrPostings`?

Reply:
Hm.. This will bubble up eventually after expand operation. `ErrPostings` is just a postings with `Next = false` and some `Err`. I think it's fine for now, we can check for errors early if that will cause problems (it will be hard to find cause of error) .